src/core/attribution/aggregation.js

```diff
@@ -25,11 +30,129 @@ export type ConnectionAggregation = {|
   +nodeAggregations: $ReadOnlyArray<NodeAggregation>,
 |};
 
+function sumScoredConnections(xs: $ReadOnlyArray<ScoredConnection>): number {
+  const reduce = (totalScore, scoredConnection) =>
+    scoredConnection.connectionScore + totalScore;
+  return xs.reduce(reduce, 0);
+}
+
+/* The correctness of this function depends on the following condition: For
+ * every ScoredConnection in xs, there is exactly one NodePrefix for which the
+ * connection's source has that NodePrefix as a prefix.  When the condition is
+ * not met, the function will almost always error, but in certain cases it can
+ * return incorrect results (example in test code).
+ */
+export function aggregateByNodePrefix(
+  xs: $ReadOnlyArray<ScoredConnection>,
+  nodePrefixes: $ReadOnlyArray<NodeAddressT>
+): $ReadOnlyArray<NodeAggregation> {
+  let connectionsFound = 0;
+  const result = [];
+  for (const nodePrefix of nodePrefixes) {
+    const connections = xs.filter(({source}) =>
```

Comment:
If you wanted to check the precondition listed in the contract, it seems
like it’d be quite simple to do: add a `const seen: Set<mixed>` at outer
level, and for each source of `connections` add the source to `visited`,
erroring if it already exists. You can rely on physical equality here
because you’re pulling objects from the same data structure, not
constructing them anew. I’d support this change if you’re willing to
make it. Then, you can remove the necessary-but-not-sufficient check
below.

Reply:
Found an alternate solution by using the new Trie class